### ISSUE TITLE
Receiving error intermittently when using cancelAll()

### ISSUE THREAD

**Issue by AzureCloudDev** [FineUploader/fine-uploader#1533]
1. BUG (seems to be)

2. Azure in basic mode

3. FineUploader v5.5.0

4. Using Firefox

5. Windows 8.1 Pro

6. Occurs intermittently. Can't reproduce all the time, please see the detailed explanation below.

7.
> onComplete: function(id, fileName, response) {
>                 var imageCount = $('#imagesTable tr td').length;
> 
>                 if (imageCount === 8) {
>                     imageUploader.cancelAll();
>                     CustomValidationSummary.addMessage('Please upload no more than 8 images.', CustomValidationSummary.MessageTypes.Error);
>                 } else {
>                     var imageId = GetPropertyValue(response, "imageId");
>                     appendUploadedImage(imageId);
>                 }
>             }
> 

I am working on functionality to try to limit the number of images a user can upload at one time. Images are uploaded, then each image is rendered inside its own table-cell, and then appended to a table-row. A user can delete the images and then upload new images again. But caveat is that there can never be more than 8 images displayed on the screen.

So the way I am trying to solve this is that upon a successful image upload, I count the number of table-cells inside the table-row. If the number of table-cells is 8, I want to cancel all remaining uploads because the user has already uploaded 8 images. Otherwise I create a new table-cell with an image and append it to the table-row. My onComplete() method is above.

This works pretty well most of the time. Yet after playing around with this for some time I noticed that intermittently I would get an error:

> TypeError: c._getFileState(...) is undefined
> ...function(a,b,d,e){var f=null==b?-1:b,g=c._getFileState(a).temp;return g.xhrs=g.x...
> 
> azure.f....min.js (line 18, col 7400)

Sometimes it would happen and sometimes it wouldn't. Sometimes it happens when I try to upload 10 images instead of 8 from the start. Sometimes it would happen when there are already 8 images on the screen and I try to delete 2, an then upload 10, then it can happen. Other times both situations fly without any errors. So I can't say that a certain condition causes it for certain. Just trying to figure out why this happens?

p.s. maxConnections: 1

Thanks.

**Comment by dbeg** [comment]
FU supports an [`itemLimit`](http://docs.fineuploader.com/branch/master/api/options.html#validation.itemLimit) within its core options. Can you leverage this functionality instead?

**Comment by AzureCloudDev** [comment]
Hi there, no I cannot. Because the user can upload images, then delete images, then upload more images, all in a single session. itemLimit wouldn't work because if a user uploaded 8 images, then deleted, 6, they can upload 6 more again, as long as there are 8 images visible on the page. itemLimit would not allow that.

**Comment by dbeg** [comment]
Alright, knowing that requirement, I suppose the first thing I would do it actually leverage the [`onSubmit`](http://docs.fineuploader.com/branch/master/api/events.html#submit) callback instead of `onComplete`, as it seems much more appropriate to validate _before_ any actual uploading has started (I suspect the error you're referencing might be occurring due to the upload actually succeeding and then some unexpected workflow being triggered thereafter, though that's just a hunch).

In any event, from the `onSubmit` validation callback: 

> Return `false` to prevent submission to the uploader.

**Comment by AzureCloudDev** [comment]
Good point! I am already doing that. If there are already 8 images uploaded then onSubmit() prevents that by returning false. However, if there are less than 8, then onSubmit doesn't catch it. In my case onComplete() awaits for server response because there is some processing that needs to be done on the image like adding it to database and scheduling it for processing, so onSubmit() fires and fails validation check before onComplete() returns, causing another upload when there has already been 8 images uploaded (just onComplete() hasn't returned yet.) so that's why I have to do this in both places. What's interesting about the error is that it only happens sometimes..seems like some internal states get out of sync or something.

**Comment by rnicholus** [comment]
Your report includes an error in the console, but what it the _actual_ issue here? What specific problem are you seeing here?

It's unclear why you can't simply use the `itemLimit` option, as this _should_ accurately reflect the non-deleted files successfully uploaded in that session.

**Comment by AzureCloudDev** [comment]
Well the issue is basically the error that's happening. It doesn't break the functionality of the FineUploader...but I am not left with a comfortable feeling of seeing javascript errors on the page. It might be a bug...thought I'd let you know..

Ok things might have been a little confusing the way I explained it, so let me try to re-explain it.
Here's how the page works: User can upload the image at which point it is shown on the page with a "Delete" button below it. So the user can upload and delete images after they have been uploaded. The requirement is to have no more than 8 images on the screen. So a user can upload 2 images, then delete them, then upload 8 images and not delete any and save everything. At this point the user would have uploaded 10 images in a current session. itemLimit would not allow for that. The reason why a user can upload and then delete images, is just for convenience purposes, just in case a user uploaded image(s) they didn't like or don't quite fit with the rest of the images, they could simply delete these image(s) and re-upload new ones, and then after everything looks good save everything on the page. Is this a little more clear now?

**Comment by rnicholus** [comment]
Again, `itemLimit` _does_ correctly account for deleted files and not count them towards the limit. So, it seems like `itemLimit` is the proper solution here.

**Comment by AzureCloudDev** [comment]
In my case the user uploads the images first, so the itemLimit is increased everytime an upload happens. Only after the image is uploaded, it is shown along with its correspinding "Delete" button. Now a user can delete the image, but deleting the image does not subtract from itemLimit, because that's internal to FineUploader. So if I upload 8 images from the start, itemLimit will be at 8, now the user deletes the images, the itemLimit is still 8. Because the user has just deleted 8 images they can upload 8 images again (requirement is to have 8 images displayed on page), but the itemLimit is already 8, effectively blocking them from doing any more uploads.

**Comment by rnicholus** [comment]
I think you're misunderstanding how the `itemLimit` validation option works. It's quite simple:

1. `itemLimit` initialized to 8
2. User uploads 8 images. Validation prevents any more from being uploaded.
3. User deletes 2 images. 2 more images may be uploaded.

This will all work as described assuming you are using Fine Uploader's delete file feature.

**Comment by AzureCloudDev** [comment]
Well that's caveat, I am not using Fine Uploader's delete feature.

**Comment by rnicholus** [comment]
Why not?

**Comment by AzureCloudDev** [comment]
I was a little more concerned about the error that sometimes occurs, and as far as I can tell I don't think I am doing anything "illegal" in the code. I am just afraid that this error could somehow break the rest of JS on the page and that would be real bad..

**Comment by AzureCloudDev** [comment]
Well, I guess I just don't, the Delete that I am using is just a html button on the page that triggers so server side functionality. I can take a look at Fine Uploader's delete, if that's the only way of doing it.

**Comment by rnicholus** [comment]
You're working against the library by deleting an uploaded file without using the built in delete file feature. Keep it simple and work _with_ the library. This is a general suggestion, applicable outside of Fine Uploader. I haven't myself seen any issues with the `cancelAll` API method. But even if there are in this scenario, it's a scenario that is overly complex and can be replaced by utilizing built-in features that are meant to support the very workflow you have described above. 

If you run into any issues with `itemLimit` or the delete file feature, please let me know. Questions regarding proper use should be posted on Stack Overflow under the fine-uploader tag. Thanks!

**Comment by AzureCloudDev** [comment]
Ray, let me ask you this, just before you close this, with the itemLimit, what will happen if a user comes to a page from a fresh page load and has already 7 images uploaded. Now he selects 8 images (when he's only really allowed to upload 1) and tries to upload. I don't think itemLimit would catch it (because it's going to think that this is a new session that allows up to 8 images to upload, but it is not)and the images would start uploading way past the total of 8 allowed, this is the scenario that I am trying to prevent.

**Comment by rnicholus** [comment]
Before initializing Fine Uploader, make a call to your server to determine the proper `itemLimit` for that use (or provide this value on page load from your server), Then, use this value when constructing a new Fine Uploader instance.

**Comment by AzureCloudDev** [comment]
I could do that, but even then, the user can delete 4 images, and have allowance to upload 4 more, so I would have to adjust the itemLimit yet again?

**Comment by rnicholus** [comment]
In that case, make use of Fine Uploader [initial file list feature](http://docs.fineuploader.com/branch/master/features/session.html) to display the previously uploaded files, with thumbnails (if desired). You can delete these files through Fine Uploader's delete files feature, and the `itemLimit` will be tracked as expected.

**Comment by AzureCloudDev** [comment]
yeah, but this would require a tighter coupling with Fine Uploader's functionality that I don't want to and don't really need to do. I just wish cancelAll() worked as expected without throwing intermittent errors. This discussion sounds like inability or inflexibility of support to investigate the issue at hand by simply providing other ways to solve the problem in arguably more complex ways.

**Comment by rnicholus** [comment]
It sounds like you want Fine Uploader to support your workflow without using built-in support for that workflow, and it's not clear why. Your current approach is at odds with the features already provided by the library to suit your requirements. Of course, you don't have to use these features, but you run the risk of exposing issues associated with what I would describe as an unsupported or edge-case use of the library. I'm sorry I don't have an answer that pleases you here, but the reality is that there is a much better and more standard way to solve support your requirements through heavily request features that a good number of users are already utilizing to solve the same exact problems. I would suggest you seriously consider updating your app to make use of these features to make your life easier,

**Comment by AzureCloudDev** [comment]
I will try to investigate the suggestion you have provided. But let me tell you, from a customer's perspective, the level of courtesy that I have received in regards to this issue, is unsatisfactory. Support hasn't even attempted to reproduce the issue, but simply stated that "they have not had any issue with the api call", that's not how support should resolve possible bug issues reported by customers. Besides it almost makes a customer feel like a liar. 
Here's how I hear it coming from support "We are not going to take a look at this even if this is a bug, try some other functionality that seems to work in most cases, oh and by the way you should have posted this on stackoverflow" Is this a normal workflow for investigating errors originating from documented api functions calls?
I think the approach I am describing here is the most simplistic of all, it requires no tight integration with Fine Uploader and requires no additional server calls to pull any data from the server. Doing it the suggested way almost feels like you have to use certain workflows to avoid running into other unfixed or new errors, which almost makes the customer wonder about the quality of the product.

**Comment by rnicholus** [comment]
Addressing your concerns:

>the level of courtesy that I have received in regards to this issue, is unsatisfactory

I have given this issue a significant amount of attention, responded to your comments promptly, and have provided guidance as to how to properly support your requirements using built-in features. 

>Support hasn't even attempted to reproduce the issue

Quoting from a previous comment of mine "I haven't myself seen any issues with the cancelAll API method".


>you should have posted this on stackoverflow

I mentioned that any further questions that are not bug reports should be posted on Stack Overflow.

>We are not going to take a look at this even if this is a bug, try some other functionality that seems to work in most cases.

I do not classify what you are seeing as a bug, as your use of the library appears to be non-standard and this is easily remedied by making use of built-in features described above.

>I think the approach I am describing here is the most simplistic of all, it requires no tight integration with Fine Uploader and requires no additional server calls to pull any data from the server.

While you may see your approach as ideal and simplistic, I see it as overly complex, brittle, and a bit confusing. By making use of feature built into the library (which you have already tied your application to) you can likely avoid all of this.

I'm sorry I wasn't able to resolve this issue to your satisfaction, but I am unfortunately not seeing any issues with this API call, and have recommended a more standard approach for you to follow that may give you better results going forward. It seems like we are going in circles at this point, so I'm going to close this case for now. Please feel free to open up a new issue if you have any bugs to report, or use Stack Overflow for questions regarding use of the library.
